src/datamodel/quests/quest_new.rs

```diff
@@ -1,64 +1,45 @@
-use std::collections::HashMap;
-use std::rc::Rc;
+use std::hash::Hash;
 
-#[derive(Debug, PartialEq)]
+use crate::core::election::Election;
+
+#[derive(Hash, PartialEq, Eq, Clone, Debug, Copy)]
 pub enum Vote {
     Success,
     Failed,
 }
 
-pub struct QuestObjective {
-    pub num_players: u32,
-    pub required_successes: u32,
-}
-
-pub struct QuestNew {
-    members: HashMap<&'static str, bool>,
-    objective: Rc<QuestObjective>,
-    success_votes: u32,
+pub struct QuestNew<'a> {
+    election: Election<Vote>,
+    winner_rule: &'a (Vote, usize),
 }
 
-impl QuestNew {
-    pub fn new(members: Vec<&'static str>, objective: Rc<QuestObjective>) -> QuestNew {
-        let members: HashMap<&'static str, bool> =
-            members.into_iter().map(|m| (m, false)).collect();
-        QuestNew {
-            objective,
-            members,
-            success_votes: 0,
-        }
+impl<'a> QuestNew<'a> {
+    pub fn new(quest_member: &[String], winner_rule: &'a (Vote, usize)) -> QuestNew<'a> {
+        assert!(quest_member.len() >= winner_rule.1, "Winner's rule cannot be bigger than quest member number");
+        let election = Election::<Vote>::new(quest_member);
+        QuestNew { election, winner_rule, }
     }
 
-    pub fn add_vote(&mut self, player: &str, vote: Vote) {
-        if self.members[player] == true {
-            return;
-        }
-        if vote == Vote::Success {
-            self.success_votes += 1;
-        }
-        *self.members.get_mut(player).unwrap() = true;
+    pub fn vote(&mut self, quest_member: &String, vote: Vote) {
+        self.election.vote(quest_member, vote);
     }
 
-    pub fn finish_quest(self) -> Result<FinishedQuestNew, QuestNew> {
-        let unvoted: Vec<bool> = self.members.iter().map(|m| *m.1).filter(|v| !*v).collect();
-        if unvoted.len() > 0 {
-            Err(self)
-        } else {
-            if self.success_votes == self.objective.required_successes {
-                Ok(FinishedQuestNew { success: true })
-            } else {
-                Ok(FinishedQuestNew { success: false })
-            }
+    pub fn finish_quest(self) -> QuestResult {
+        let scrutiny = self.election.count_votes().ok().unwrap();
+        match (self.winner_rule.0 , scrutiny.result().get(&Some(self.winner_rule.0))) { 
```

Comment:
I think we should have both fields separate with their own names, or have a `WinnerRule` class with a method. I don't know by looking at this code what `winner_rule.0` means. In general tuples should have a clear meaning of their contents by their name.

Reply:
ok, i'll add a `WinnerRule` `struct` to make it more readable.